### ISSUE TITLE
Support importing existing infrastructure

### ISSUE THREAD

**Issue by hasanaburayyan** [winglang/wing#5170]
### Feature Spec

Wing resources can now be imported into your application code. This enables users to interact with resources that may have been created outside of the given Wing project. Whereas before Wing applications were expected to be entirely greenfield, now its easy to write poc, or one off jobs in Wing that interact with existing infrastructure in your cloud environment. 

### Use Cases

Import existing buckets, tables, redis, and other resources into your Wing application without having to provision them as part of your deployment process.

### Implementation Notes

One thought on implementation might be that resources in Wing have an optional `ImportOptions` property that contain any required fields that are needed to perform the import for the given Platform.

(Rough sketch maybe)
```js
bring cloud;

let b = new cloud.Bucket(import: {name: "my-existing-bucket"})
```

Related Slack thread: https://winglang.slack.com/archives/G04G3TCHE12/p1701893450410499

### Community Notes

* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.
* If this issue is labeled **needs-discussion**, it means the spec has not been finalized yet. Please reach out on the #dev channel in the [Wing Slack](https://t.winglang.io/slack).

**Comment by eladb** [comment]
Another thing to consider is what's the `sim` behavior of these resources. It could be:

1. Use a local simulated bucket.
2. Reference an S3 bucket in the cloud set up for development.

**Comment by ekeren** [comment]
> let bucket = aws.Bucket.fromArn("arn:....");

We've just heard from a team deploying "micro-stacks" that are sharing resources on the network level and data storage level (like RDS, Shared S3 etc). Each stack has a different TF State file stored on S3.

It feels like this can be a common use case, and we should support it on the `cloud` namespace level. 
Something like:
```
let bucket = cloud.Bucket.from(j: Json) 
```

And the platform provider, using values or inputs ([rfc](https://github.com/winglang/wing/pull/5171)), is responsible to produce this Json, in the case of AWS it is probably `{ arn: '....' }`

**Comment by eladb** [comment]
> we should support it on the `cloud` namespace level

Why does it matter to support this at the `cloud` abstract classes instead of the concrete resources?

**Comment by ekeren** [comment]
> Why does it matter to support this at the cloud abstract classes instead of the concrete resources?

I don't think that the multi micro-stack applications that share a bucket means that my project needs to be bound to a specific cloud provider, the project can still be cloud agnostic...

**Comment by eladb** [comment]
>  I don't think that the multi micro-stack applications that share a bucket means that my project needs to be bound to a specific cloud provider, the project can still be cloud agnostic...

Got it. So the use case is to import an existing resource from another *Wing* application. Correct?

**Comment by ekeren** [comment]
> Got it. So the use case is to import an existing resource from another Wing application. Correct?

I am not sure. 

One counter example is: when I choose to migrate from TF to Wing, micro-stack after another, the existing resource will not be from an existing another Wing application till the migration is over...

**Comment by eladb** [comment]
I suspect that until the resource is "elevated" into Wing, it is just an `aws.Bucket`. Once it is "elevated", it becomes a `cloud.Bucket`. No?

**Comment by ekeren** [comment]
I think that this temporary state can take decades... I would consider this as a stable state.

**Comment by eladb** [comment]
OK, so this only reinforces the fact that in those cases it would make more sense to import the `aws.Bucket` instead of the `cloud.Bucket` because the app is basically depending on a real S3 bucket in the real reality of the real world.

**Comment by marciocadev** [comment]
I have a question about the implementation, is it already possible to create static methods?

I ask this because in the past, I had to implement the Schedule by creating a schedule and then creating the cron instead of creating the cron through a static method of the Schedule class.

An implementation to obtain an existing resource "requires" the use of a static method, and I don't know if the compiler interprets that.

**Comment by hasanaburayyan** [comment]
@marciocadev static methods are supported now, so that should be fine. 

Maybe Im missing something @eladb and @ekeren but I think even with `aws.Bucket` we would have an issue of ambiguity between provisioning engines for the preflight API.

I.E.
```js
bring aws;

let b = aws.Bucket.fromArn("some arn");

// b is typed as cloud.Bucket
// thus we have access to some preflight APIs such as "addFile" as such
b.addFile("myfile.txt", "./myfile.txt");
```

This means that we need to have the imported bucket creation go through our polycon system and do something something like `App.of(this).newAbstract("@winglang/sdk.cloud.BucketImport")` (something like this)

Allowing platforms to decide how to import a bucket. I.E. since `awscdk` would use `aws_s3_Bucket.from..` and `cdktf` would use a data lookup. 

I think this is why I was in favor of an api where we pass import options at the `cloud.Bucket` level, but trying to manage interfaces for import options could become a nightmare.

**Comment by eladb** [comment]
When using a referenced resource, users reference it from an well-known environment, which means that I don't think there is really a lot of value for adding another layer of decoupling (between `tf-aws` and `awscdk` in this case).

This implies that maybe we just need to separate the `aws` module to `bring tfaws` and `bring awscdk` and each of them will have their own `tfaws.Bucket` (there could be some implementation reuse under the hood but that's it):

```js
bring tfaws;

let b = tfaws.Bucket.fromArn(...);
```

**Comment by hasanaburayyan** [comment]
> which means that I don't think there is really a lot of value for adding another layer of decoupling (between tf-aws and awscdk in this case).

Okay I think it makes sense to go on this hypothesis

**Comment by Chriscbr** [comment]
Reading the initial issue description, it's not clear to me whether the behavior of this feature is to allow resources to be **referenced** or to be **imported**.

When someone uses a tool like Terraform or CloudFormation to manage their infrastructure, **importing a resource** refers to taking an existing cloud resource that exists a resource provider (for example, an S3 bucket you created in the AWS console), and thereafter managing the resource's state with the IaC tool. Once an import has been completed, you can update resource attributes or delete the resource through the IaC tool.

In AWS CDK, two kinds of operations were supporting -- referencing resources and importing them:
- Methods like `Bucket.fromBucketName()` allow the user to **reference external resources**. This lets other resources refer to the resource's static information, even augment it in some cases (e.g. create a new bucket event handler, or grant a lambda function access to the bucket). But the resource's main attributes cannot be modified. Docs: https://docs.aws.amazon.com/cdk/v2/guide/resources.html#resources_referencing
- `cdk import` is a capability that allows resources to be **imported** and thereafter managed by the application's IaC state. This requires more manual effort since you may need to model the state in your Terraform/Cloudformation/CDK app first, double check that all of the attributes match (i.e. you haven't introduced drift), and then run a CLI to command to finish the operation. Docs: https://aws.amazon.com/blogs/devops/how-to-import-existing-resources-into-aws-cdk-stacks/
  - Terraform supports this through the `terraform import` command

**Comment by hasanaburayyan** [comment]
> it's not clear to me whether the behavior of this feature is to allow resources to be referenced or to be imported.

This is a good distinction Chris! I have just been speaking with the implications that this if for the "reference" capability. I think for "importing state management" is a separate can of worms. 

Also (based on our discussion) this issue is probably blocked by: https://github.com/winglang/wing/issues/5526 if we want to support static methods.

**Comment by revitalbarletz** [comment]
@hasanaburayyan what is the status of this one?

**Comment by Chriscbr** [comment]
Noting an idea from @MarkMcCulloh - maybe in the future there can be some kind of syntax sugar for these "references" to existing infrastructure once we've settled on some pattern or idiom for them, like `&Bucket` etc. (Maybe there could be semantic implications to the syntax, like whether you can call mutating preflight methods)

**Comment by eladb** [comment]
Nice!
